lib/tx_build/preconditions.ex

```diff
@@ -0,0 +1,185 @@
+
+  @type t :: %__MODULE__{type: type(), preconditions: preconditions()}
+
+  defstruct [:type, :preconditions]
+
+  @impl true
+  def new(preconditions \\ [], opts \\ [])
+
+  def new([], _opts) do
+    %__MODULE__{type: :none, preconditions: nil}
+  end
+
+  def new([time_bounds: %TimeBounds{} = time_bounds], _opts) do
+    %__MODULE__{type: :precond_time, preconditions: time_bounds}
+  end
+
+  def new(preconditions, _opts) do
+    time_bounds = Keyword.get(preconditions, :time_bounds)
+    ledger_bounds = Keyword.get(preconditions, :ledger_bounds)
+    min_seq_num = Keyword.get(preconditions, :min_seq_num)
+    min_seq_age = Keyword.get(preconditions, :min_seq_age, 0)
+    min_seq_ledger_gap = Keyword.get(preconditions, :min_seq_ledger_gap, 0)
+    extra_signers = Keyword.get(preconditions, :extra_signers)
+
+    with {:ok, time_bounds} <- validate_time_bounds(time_bounds),
+         {:ok, ledger_bounds} <- validate_ledger_bounds(ledger_bounds),
+         {:ok, min_seq_num} <- validate_min_seq_num(min_seq_num),
+         {:ok, min_seq_age} <- validate_min_seq_age(min_seq_age),
+         {:ok, min_seq_ledger_gap} <- validate_min_seq_ledger_gap(min_seq_ledger_gap),
+         {:ok, extra_signers} <- validate_extra_signers([], extra_signers) do
+      %__MODULE__{
+        type: :precond_v2,
+        preconditions: %{
+          time_bounds: time_bounds,
+          ledger_bounds: ledger_bounds,
+          min_seq_num: min_seq_num,
+          min_seq_age: min_seq_age,
+          min_seq_ledger_gap: min_seq_ledger_gap,
+          extra_signers: extra_signers
+        }
+      }
+    end
+  end
+
+  @impl true
+  def to_xdr(%__MODULE__{type: :none, preconditions: nil}) do
+    preconditions_type = PreconditionType.new(:PRECOND_NONE)
+
+    Preconditions.new(Void.new(), preconditions_type)
+  end
+
+  def to_xdr(%__MODULE__{
+        type: :precond_time,
+        preconditions: %TimeBounds{min_time: min_time, max_time: max_time}
+      }) do
+    preconditions_type = PreconditionType.new(:PRECOND_TIME)
+
+    min_time = TimePoint.new(min_time)
+    max_time = TimePoint.new(max_time)
+
+    min_time
+    |> TimeBoundsXDR.new(max_time)
+    |> Preconditions.new(preconditions_type)
+  end
+
+  def to_xdr(%__MODULE__{
+        type: :precond_v2,
+        preconditions: %{
+          time_bounds: time_bounds,
+          ledger_bounds: ledger_bounds,
+          min_seq_num: min_seq_num,
+          min_seq_age: min_seq_age,
+          min_seq_ledger_gap: min_seq_ledger_gap,
+          extra_signers: extra_signers
+        }
+      }) do
+    preconditions_type = PreconditionType.new(:PRECOND_V2)
+
+    time_bounds = TimeBounds.to_xdr(time_bounds)
+    ledger_bounds = LedgerBounds.to_xdr(ledger_bounds)
+    min_seq_num = min_seq_num |> SequenceNumber.to_xdr() |> OptionalSequenceNumber.new()
+    min_seq_age = Duration.new(min_seq_age)
+    min_seq_ledger_gap = UInt32.new(min_seq_ledger_gap)
+
+    extra_signers =
```

Comment:
@CristhianRodriguezMolina @FelipeGuzmanSierra You can use a function to perform this operation.